### PR TITLE
add shot-noise to measurement

### PR DESCRIPTION
**Pull request by multiphaseCFD**
This PR adds support for finite-shot measurement statistics (`expval`, `var`, and `probs`) for `lightning.qubit` and `lightning.kokkos` devices.
 

[sc-46423]

### CODE REVIEW

File: runtime/include/QuantumDevice.hpp
Comment (maliasadi) [reply runtime/include/QuantumDevice.hpp]: I believe @multiphaseCFD followed the signature in PL (check [here](https://github.com/PennyLaneAI/pennylane/blob/a05ae286c5925e0c881099cea751193d368779fc/pennylane/measurements/measurements.py#L490)). But at the moment, it doesn't have any use cases in Lightning and Catalyst. If you pass an empty list for the range of samples to use, aa samples are going to be used by default. I took this into account when adding the support in Catalyst.

Comment (multiphaseCFD) [reply runtime/include/QuantumDevice.hpp]: Just as @maliasadi mentioned above. I followed the signature in PL

File: frontend/test/pytest/test_measurements_results.py
Comment (maliasadi) [inline frontend/test/pytest/test_measurements_results.py:201]: @multiphaseCFD as discussed offline, there are some inconsistent error messages raising for 'lightning.qubit' and 'lightning.kokkos' for `expval`, `var`, and `probs`.

File: runtime/tests/Test_LightningMeasures.cpp
Comment (maliasadi) [inline runtime/tests/Test_LightningMeasures.cpp:1030]: @multiphaseCFD this is related to the Lightning bug that we discussed. We should uncomment this after merging the fix.

File: runtime/tests/Test_LightningMeasures.cpp
Comment (maliasadi) [inline runtime/tests/Test_LightningMeasures.cpp:1076]: I've added one more test with the same issue while working on the support.. this should be uncommented before merge too.

File: doc/changelog.md
Comment (maliasadi) [inline doc/changelog.md:96]: Happy your first contribution to Catalyst! :tada:

File: frontend/test/pytest/test_measurements_results.py
Comment (maliasadi) [reply frontend/test/pytest/test_measurements_results.py:201]: Our current CI failures are related to this issue.

File: runtime/include/QuantumDevice.hpp
Comment (dime10) [reply runtime/include/QuantumDevice.hpp]: Thanks! Still curious how someone would use it, but that's unrelated to the PR 😅

File: frontend/test/pytest/test_measurements_results.py
Comment (multiphaseCFD) [reply frontend/test/pytest/test_measurements_results.py:201]: How about setting the error message as `xxx observables do not have shots support.`?

File: doc/dev/quick_start.rst
Comment (josh146) [inline doc/dev/quick_start.rst:257]: Does this unlock finite shots for probs as well?

File: frontend/test/pytest/test_measurements_results.py
Comment (josh146) [inline frontend/test/pytest/test_measurements_results.py:203]: I'm curious how come tests for expval/var with non `qml.Hermitian` observables aren't added?

Comment (josh146) [reply frontend/test/pytest/test_measurements_results.py:203]: I wonder if it is worth porting over some of the finite-shot measurement tests from here? https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/devices/tests/test_measurements.py#L198

**Comment by josh146** [comment]
This is great @multiphaseCFD @maliasadi, great work! 

Something I do wonder about though is testing. We know from experience that there are a lot of edge cases for measurements, such as:

- Single observables
- tensors on different wires
- linear combinations/hamiltonians
- new operator arithmetic
- differentiation
- expval/var

Should we have more extensive testing here?

File: frontend/test/pytest/test_measurements_results.py
Comment (maliasadi) [reply frontend/test/pytest/test_measurements_results.py:203]: We had tests with other observables and shots > 0 in the frontend (both in `test_measurements_results.py` and `test_pytree_args.py`). To pass those tests, I only had to increase the number of shots :) 

 Thanks for the link. I'll add more tests :+1:

File: doc/dev/quick_start.rst
Comment (maliasadi) [reply doc/dev/quick_start.rst:257]: yes, I've added it to README of the runtime. But, we didn't have any mention of this unsupported feature for `probs` in the quick start demo, so didn't update anything here.

File: frontend/test/pytest/test_measurements_results.py
Comment (maliasadi) [reply frontend/test/pytest/test_measurements_results.py]: @josh146 added a new test file for testing these finite-shots features. [54489e5](https://github.com/PennyLaneAI/catalyst/pull/392/commits/54489e5e25745c2003c81acd7470fd9b2c2799e3) those commented tests are related to the bug I mentioned this morning and @multiphaseCFD is currently working on it.

File: doc/dev/quick_start.rst
Comment (dime10) [inline doc/dev/quick_start.rst:252]: Is this more accurate?
```suggestion
     - The expectation value of all observables is supported analytically, as well as with finite-shots except for Hermitian.
```

File: frontend/test/pytest/test_measurements_results.py
Comment (dime10) [inline frontend/test/pytest/test_measurements_results.py:630]: Why so many shots? 😅 (not sure if it impacts the runtime or not)

File: frontend/test/pytest/test_pytree_args.py
Comment (dime10) [inline frontend/test/pytest/test_pytree_args.py:222]: Is there a reason to convert an existing test from analytic to finite-shots?

File: runtime/README.rst
Comment (dime10) [inline runtime/README.rst:82]: Do we not support observables for probs now?

**Review by dime10 — COMMENTED** [review]
Nice work 👍

File: frontend/test/pytest/test_measurements_shots_results.py
Comment (josh146) [inline frontend/test/pytest/test_measurements_shots_results.py]: How come this one is commented out? If the test doesn't work, might be best to use 
```python
@pytest.mark.xfail(reason="...")
```

so we don't lose track of this test

File: frontend/test/pytest/test_measurements_results.py
Comment (josh146) [reply frontend/test/pytest/test_measurements_results.py]: Thanks Ali, this is great! very thorough

File: frontend/test/pytest/test_measurements_shots_results.py
Comment (maliasadi) [reply frontend/test/pytest/test_measurements_shots_results.py]: It's uncommented now! We got the support working for all observables (except `Hermitian`) :)

File: frontend/test/pytest/test_measurements_results.py
Comment (maliasadi) [reply frontend/test/pytest/test_measurements_results.py]: Fixed! :smile:

File: doc/dev/quick_start.rst
Comment (maliasadi) [reply doc/dev/quick_start.rst:252]: This is not true for all backend devices though. I added the details for each device to `runtime/README`.

File: .github/workflows/build-wheel-linux-x86_64.yaml
Comment (maliasadi) [reply .github/workflows/build-wheel-linux-x86_64.yaml:262]: This is temporary.. I'll fix this after merging https://github.com/PennyLaneAI/pennylane-lightning/pull/577

File: frontend/test/pytest/test_pytree_args.py
Comment (maliasadi) [reply frontend/test/pytest/test_pytree_args.py]: Fixed!

File: runtime/README.rst
Comment (maliasadi) [reply runtime/README.rst]: Observable support for `probs`, `sample`, and `counts` is a different ticket (wasn't P0 for this quarter). But, Shuli and I had worked on this support too. This is now supported for all of Lightning sims, thanks to Shuli's hard work! We can tackle it next quarter as P0.

Comment (dime10) [reply runtime/README.rst]: Yeah I remember seeing it added to lightning. I guess it can be a different PR on the Catalyst if it's too much work.

File: doc/dev/quick_start.rst
Comment (dime10) [reply doc/dev/quick_start.rst:252]: I've updated the suggestion, how does that sound? The current formulation doesn't read that well.

File: runtime/README.rst
Comment (maliasadi) [reply runtime/README.rst]: It's not too much work, but it requires changes into the Quantum dialect and lowering passes as well as the runtime. My plan for this ticket was helping Shuli to get a better understanding of the runtime :) I didn't want to implement this support originally but the situation changed in the last min.. hehe.

Comment (maliasadi) [reply runtime/README.rst]: Implementing the support is a half-day work, I can give it a try next week and I hope we don't run into any bugs.

File: doc/dev/quick_start.rst
Comment (maliasadi) [reply doc/dev/quick_start.rst:252]: Sounds good! I'll update it in other places too. Thanks :+1:

File: .github/workflows/build-wheel-linux-x86_64.yaml
Comment (maliasadi) [reply .github/workflows/build-wheel-linux-x86_64.yaml:262]: yes, I added this tag before finding some issues (resolved in PR 577). I'll fix the tag as soon as we get that PR merged.

Comment (dime10) [reply .github/workflows/build-wheel-linux-x86_64.yaml:262]: Ok so the intention is to not merge this until that PR is merged?

Comment (multiphaseCFD) [reply .github/workflows/build-wheel-linux-x86_64.yaml:262]: The PR was merged! @maliasadi

Comment (maliasadi) [reply .github/workflows/build-wheel-linux-x86_64.yaml]: Done!

File: .github/workflows/build-wheel-linux-x86_64.yaml
Comment (maliasadi) [inline .github/workflows/build-wheel-linux-x86_64.yaml:262]: @dime10 this is the latest commit including all fixes. Are you happy with this or you prefer to tag the "master" branch for wheels?

File: doc/changelog.md
Comment (maliasadi) [inline doc/changelog.md:104]:
```suggestion
Erick Ochoa Lopez,
Shuli Shu.
```

File: frontend/test/pytest/test_measurements_shots_results.py
Comment (multiphaseCFD) [reply frontend/test/pytest/test_measurements_shots_results.py]: Hey @josh146,  To get `Hermitian` support, we need get `LAPACK` support to obtain `eigenvalues` and `unitary` of a `Hermitian` matrix. As discussed with @mlxd, it will be an issue to link against `LAPACK` on `Windows` systems. So I started with Linux systems and created a PR [569](https://github.com/PennyLaneAI/pennylane-lightning/pull/569). It works fine on Perlmutter with pre-install `LAPACK` lib. We still need some tests for other Linux system. Per `Windows` support, I plan to use `ExternalProject_Add` function [here](https://cmake.org/cmake/help/latest/module/ExternalProject.html) to build `LAPACK` if `LAPACK` is not found after I'm back from vacation. Will keep you posted on the progress. BTW, this [story](https://app.shortcut.com/xanaduai/story/51606/add-hermitian-obs-support-for-measurement-with-shot-noise) was created to track the progress on the `Hermitian` support.

File: .github/workflows/build-wheel-linux-x86_64.yaml
Comment (dime10) [reply .github/workflows/build-wheel-linux-x86_64.yaml:262]: The specific commit is good, for master we have the plugin test matrix :)

**Review by dime10 — APPROVED** [review]
🚀

**Review by maliasadi — COMMENTED** [review]
Thanks @multiphaseCFD! :100:

**Pull request merged**
